File: src/main/java/dao/Dao.java
package dao;

import javax.persistence.EntityManager;
import java.util.List;

public interface Dao<T> {

	EntityManager getEntityManager();

	List<T> listarTodos();

	T salvar(T entity);

	T atualizar(T entity);

	T encontraPeloId(T entity);
}

File: src/main/java/dao/Dao.kt
package dao

import org.hibernate.Session
import javax.persistence.EntityManager

interface Dao <T> {
    fun getEntityManager(): EntityManager

    fun getSession(): Session

    fun listarTodos(): List<T>

    fun salvar(entity: T): T

    fun atualizar(entity: T): T

    fun encontraPeloId(entity: T): T
}